sos/plugins/etcd.py

```diff
@@ -23,15 +23,46 @@ class etcd(Plugin, RedHatPlugin):
     """etcd plugin
     """
 
+    packages = ('etcd',)
+
     def setup(self):
-        self.add_copy_spec("/etc/etcd")
+        self.address = self.get_address()
+        self.cmd = 'etcdctl '
+
+        self.add_copy_spec('/etc/etcd')
+
+        subcmds = ['--version',
+                   'member list',
+                   'cluster-health',
+                   'ls --recursive',
+                   ]
+
+        for sub in subcmds:
+            self.add_cmd_output("%s %s" % (self.cmd, sub))
+
+        api_urls = ['/v2/stats/leader',
+                    '/v2/stats/self',
+                    '/v2/stats/store',
+                    ]
+
+        for url in api_urls:
+            self.add_cmd_output("curl -s %s%s" % (self.address, url))
 
-        self.add_cmd_output("curl http://localhost:4001/version")
-        self.add_cmd_output("curl http://localhost:4001/v2/members")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/leader")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/self")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/store")
         self.add_cmd_output("ls -lR /var/lib/etcd/")
 
+    def get_address(self):
+        try:
+            with open('/etc/etcd/etcd.conf', 'r') as ef:
+                for line in ef:
+                    if line.startswith('ETCD_LISTEN_CLIENT_URLS'):
+                        return line.split('=')[1].replace('"', '').strip()
+        # If we can't read etcd.conf, assume defaults by etcd version
+        except IOError:
+            ver = self.get_command_output('etcdctl --version')['output']
+            ver = ver.split('\n')[0].split(':')[1].strip()
+            if ver.startswith('2'):
+                return 'http://localhost:4001'
+            if ver.startswith('3'):
+                return 'http://localhost:2379'
```

Comment:
There should either be a default case here (in case for whatever reason the version information we receive is garbled or missing), or the calling code needs to know how to deal with a `None` return here: right now the main `setup()` code will just carry on merrily formatting `None` into a bunch of URLs...